### PR TITLE
Separate notifications for comment replies #4316

### PR DESCRIPTION
**Pull request by MasterHans**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x ] No

**Other information:**
Allow notifications to followers participated in current comment replay and parent object owner only

### CODE REVIEW

**Comment by luke-** [comment]
Can you also create some tests for this?

**Comment by MasterHans** [comment]
> Can you also create some tests for this?

Yes, of course. I'll do it

**Comment by MasterHans** [comment]
> Can you also create some tests for this?

Done but have questions...

1) I can not fully test sending notifications to the users, because it has mail sending with queue->push
So basic test grabber $this->grabSentEmails() did not actually get sent mails.
So to check it I need to perform PHP yii queue/run command for test environment, but if I do it like this:

```
cd /var/www/humhub/protected/humhub/tests/codeception/bin
php yii queue/run
```

or  like this

`$result = shell_exec("php /var/www/humhub/protected/humhub/tests/codeception/bin/yii queue/run");`

It's did not work...

**Comment by ArchBlood** [comment]
> > Can you also create some tests for this?
> 
> Done but have questions...
> 
> 1. I can not fully test sending notifications to the users, because it has mail sending with queue->push
>    So basic test grabber $this->grabSentEmails() did not actually get sent mails.
>    So to check it I need to perform PHP yii queue/run command for test environment, but if I do it like this:
> 
> ```
> cd /var/www/humhub/protected/humhub/tests/codeception/bin
> php yii queue/run
> ```
> 
> or like this
> 
> `$result = shell_exec("php /var/www/humhub/protected/humhub/tests/codeception/bin/yii queue/run");`
> 
> It did not work...
> 
> In the test environment only one notification sending steadily. This notification is to the admin user.

Does using `proc_open()` help any better when using the command? Also depending on the system used `shell_exec()` can fail, namely Windows which in that case `popen()` is recommended by PHP

**Comment by luke-** [comment]
@buddh4 Do you have a suggestion or idea here?

**Comment by buddh4** [comment]
@MasterHans the testenvornment uses the instant queue so notifications should be send immediately see 

- Unit test mail functions: https://github.com/humhub/humhub/blob/master/protected/humhub/tests/codeception/_support/HumHubDbTestCase.php#L211
- Example https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/tests/codeception/unit/MentionTest.php#L32

File: protected/humhub/modules/comment/models/Comment.php [hunk 1]
```diff
@@ -151,6 +151,28 @@ public function afterSave($insert, $changedAttributes)
...
+            if ($this::isSubComment($this)) {
+                $parent = $this->object_model::findOne(['id', $this->object_id]);
+                $parentUser = User::findOne(['id' => $parent->created_by]);
+
+                /**
+                 * Send notifications to only these followers:
+                 * author of the parent comment
+                 * author of the commented record
+                 */
+                $followersShouldReceiveNotification = [
```
Comment (buddh4) [inline protected/humhub/modules/comment/models/Comment.php:163]: Shouldn't we add also all users already created a sub comment to the same parent?

File: protected/humhub/modules/comment/models/Comment.php [hunk 2]
```diff
@@ -300,4 +322,16 @@ public function isUpdated()
...
     }
+
+    /**
+     * Checks if given content object is a subcomment
+     *
+     * @param $object
+     * @return bool
+     * @since 1.8
+     */
+    public static function isSubComment($object)
```
Comment (buddh4) [inline protected/humhub/modules/comment/models/Comment.php:333]: We need to make sure to not collide with https://github.com/humhub/humhub/pull/4571/files#diff-4fcd54238adfe8774ee607eeadd9ff1c821166188d5f9dabd37cac4e2729a218R311

**Review by buddh4 — CHANGES_REQUESTED** [review]

File: protected/humhub/modules/comment/models/Comment.php [hunk 3]
```diff
@@ -151,6 +151,28 @@ public function afterSave($insert, $changedAttributes)
                 }, $mentionedUsers)]);
             }
 
+            if ($this::isSubComment($this)) {
+                $parent = $this->object_model::findOne(['id', $this->object_id]);
+                $parentUser = User::findOne(['id' => $parent->created_by]);
```
[the remark this answers is not in the document]
Comment (MasterHans) [reply protected/humhub/modules/comment/models/Comment.php:156]: > `$parent->createdBy` returns the user of parent comment

That is right, but we need not only id of the user but full ActiveRecord of the user. That is why I used:
`$parentUser = User::findOne(['id' => $parent->created_by]);`

[the remark this answers is not in the document]
Comment (buddh4) [reply protected/humhub/modules/comment/models/Comment.php:156]: `$parent->createdBy` returns the full record `$parent->created_by` is only the id

File: protected/humhub/modules/comment/models/Comment.php [hunk 4]
```diff
@@ -151,6 +151,28 @@ public function afterSave($insert, $changedAttributes)
                 }, $mentionedUsers)]);
             }
 
+            if ($this::isSubComment($this)) {
+                $parent = $this->object_model::findOne(['id', $this->object_id]);
```
[the remark this answers is not in the document]
Comment (MasterHans) [reply protected/humhub/modules/comment/models/Comment.php:155]: Done!

File: protected/humhub/modules/comment/models/Comment.php [hunk 5]
```diff
@@ -151,6 +151,28 @@ public function afterSave($insert, $changedAttributes)
                 }, $mentionedUsers)]);
             }
 
+            if ($this::isSubComment($this)) {
```
[the remark this answers is not in the document]
Comment (MasterHans) [reply protected/humhub/modules/comment/models/Comment.php:154]: Done other way: remove function add  `$this->object_model === Comment::class` condition instead

File: protected/humhub/modules/comment/models/Comment.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (MasterHans) [reply protected/humhub/modules/comment/models/Comment.php:156]: Done!

File: protected/humhub/modules/comment/models/Comment.php (hunk 1 above)
Comment (MasterHans) [reply protected/humhub/modules/comment/models/Comment.php:163]: Done!

File: protected/humhub/modules/comment/models/Comment.php (hunk 2 above)
Comment (MasterHans) [reply protected/humhub/modules/comment/models/Comment.php:333]: Done! Remove this method

**Comment by MasterHans** [comment]
> @MasterHans the testenvironment uses the instant queue so notifications should be send immediately see:
> 
> * Unit test mail functions: https://github.com/humhub/humhub/blob/master/protected/humhub/tests/codeception/_support/HumHubDbTestCase.php#L211
> * Example https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/tests/codeception/unit/MentionTest.php#L32

Done!

**Pull request closed without merging**